File: backend/app/schemas/collaboration.py
from datetime import datetime
from typing import Annotated, Literal

from pydantic import Field

from app.schemas.architect import LongText, ShortText
from app.schemas.artifact import ArtifactCollection, MissionArtifact
from app.schemas.base import Schema
from app.schemas.execution import ExecutionRequest, ExecutionResult
from app.schemas.memory import OrganizationMemory

ConversationMessageType = Literal[
    "suggestion",
    "question",
    "answer",
    "decision",
    "review_request",
    "review_response",
    "information",
    "warning",
]
CollaborationStage = Literal["before_execution", "during_execution", "after_execution"]


class ConversationMessage(Schema):
    """One attributable worker-to-worker collaboration record."""

    message_id: ShortText
    session_id: ShortText
    sender_worker_id: ShortText
    sender_department: ShortText
    receiver_worker_id: ShortText | None = None
    related_task_id: ShortText
    message_type: ConversationMessageType
    content: LongText
    timestamp: datetime


class WorkerConversation(Schema):
    """A phase-scoped, chronological thread of worker messages."""

    conversation_id: ShortText
    phase_id: Annotated[int, Field(ge=1, le=32)]
    phase_name: ShortText
    messages: Annotated[list[ConversationMessage], Field(min_length=1, max_length=128)]


class CollaborationSession(Schema):
    """The complete communication state for one Genesis mission."""

    session_id: ShortText
    conversations: Annotated[list[WorkerConversation], Field(max_length=32)]
    created_at: datetime
    updated_at: datetime


class CollaborationMessageDraft(Schema):
    """Provider-authored content before the Collaboration Engine assigns message metadata."""

    message_type: ConversationMessageType
    content: LongText
    receiver_worker_id: ShortText | None = None


class CollaborationContext(Schema):
    """Relevant shared knowledge supplied to a worker before task execution."""

    organization_memory: OrganizationMemory
    relevant_artifacts: Annotated[list[MissionArtifact], Field(max_length=12)]
    previous_decisions: Annotated[list[ConversationMessage], Field(max_length=12)]
    conversation_history: Annotated[list[ConversationMessage], Field(max_length=24)]


class CollaborativeExecutionRequest(Schema):
    """Existing execution data plus collaboration state; the original request is unchanged."""

    execution_request: ExecutionRequest
    artifact_collection: ArtifactCollection = Field(
        default_factory=lambda: ArtifactCollection(artifacts=[])
    )
    collaboration_session: CollaborationSession | None = None


class CollaborativeExecutionResult(Schema):
    """Execution output enriched with the updated collaboration session."""

    execution_result: ExecutionResult
    collaboration_session: CollaborationSession

File: frontend/lib/api/collaboration.ts
import type { ArtifactCollection } from "@/lib/api/artifacts";
import type { ExecutionResult } from "@/lib/api/execution";
import type { OrganizationMemory } from "@/lib/api/memory";
import type { TaskGroup } from "@/lib/api/task-generator";
import type { WorkerAssignmentResult } from "@/lib/api/worker-assignment";
import type { Workflow } from "@/lib/api/workflow";

export type ConversationMessageType =
  | "suggestion"
  | "question"
  | "answer"
  | "decision"
  | "review_request"
  | "review_response"
  | "information"
  | "warning";

export type ConversationMessage = {
  content: string;
  message_id: string;
  message_type: ConversationMessageType;
  receiver_worker_id: string | null;
  related_task_id: string;
  sender_department: string;
  sender_worker_id: string;
  session_id: string;
  timestamp: string;
};

export type WorkerConversation = {
  conversation_id: string;
  messages: ConversationMessage[];
  phase_id: number;
  phase_name: string;
};

export type CollaborationSession = {
  conversations: WorkerConversation[];
  created_at: string;
  session_id: string;
  updated_at: string;
};

export type CollaborativeExecutionResult = {
  collaboration_session: CollaborationSession;
  execution_result: ExecutionResult;
};

type ApiErrorPayload = { code?: string; message?: string };

export class CollaborationApiError extends Error {
  readonly code: string;
  readonly status: number;

  constructor({
    code,
    message,
    status,
  }: Readonly<{ code: string; message: string; status: number }>) {
    super(message);
    this.name = "CollaborationApiError";
    this.code = code;
    this.status = status;
  }
}

const apiBaseUrl = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export async function executeCollaborativeReadyTasks(
  input: Readonly<{
    artifactCollection?: ArtifactCollection;
    collaborationSession?: CollaborationSession;
    organizationMemory?: OrganizationMemory;
    taskGroups: TaskGroup[];
    workerAssignmentResult: WorkerAssignmentResult;
    workflow: Workflow;
  }>,
): Promise<CollaborativeExecutionResult> {
  let response: Response;

  try {
    response = await fetch(`${apiBaseUrl}/api/v1/collaborative-executions`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        artifact_collection: input.artifactCollection ?? { artifacts: [] },
        collaboration_session: input.collaborationSession ?? null,
        execution_request: {
          organization_memory: input.organizationMemory ?? { entries: [] },
          task_groups: input.taskGroups,
          worker_assignment_result: input.workerAssignmentResult,
          workflow: input.workflow,
        },
      }),
    });
  } catch {
    throw new CollaborationApiError({
      code: "collaboration_engine_unreachable",
      message: "Genesis could not reach the Collaboration Engine. Please try again.",
      status: 0,
    });
  }

  if (!response.ok) {
    const payload = (await response.json().catch(() => ({}))) as ApiErrorPayload;
    throw new CollaborationApiError({
      code: payload.code ?? "collaboration_engine_error",
      message: payload.message ?? "Genesis could not coordinate the organization.",
      status: response.status,
    });
  }

  return (await response.json()) as CollaborativeExecutionResult;
}
